narwhal/primary/src/state_handler.rs

```diff
@@ -447,6 +437,57 @@ impl RandomnessState {
                 return;
             }
         };
+
+        // Try to verify the aggregated signature all at once. (Should work in the happy path.)
+        if ThresholdBls12381MinSig::verify(dkg_output.vss_pk.c0(), &round.signature_message(), &sig)
+            .is_err()
+        {
+            // If verifiation fails, some of the inputs must be invalid. We have to go through
+            // one-by-one to find which.
+            self.partial_sigs.retain(|&(r, authority_id), partial_sig| {
```

Comment:
IIUC one bad signature would put us in the same performance situation as we are in now right? 

I know we haven't gotten to the bottom of exactly what is causing the perf issues, but for nw cert signature verification we used to start up separate tasks to verify signatures in parallel and so that the work was not done on the main thread. I wonder if we could benefit from doing the same here?

Reply:
Yes it would, but a bad signature would require byzantine validator, which is unlikely and we can easily detect/blame whoever is responsible.

As for threads, this work is already not done on the "main thread" with respect to narwhal round advancement - it's in the state handler loop which should not be blocking anything to do with narwhal rounds? But your test already commeted out the one thing in the vote request hadnler that is part of round advancement, and it didn't help much. hence this change on the theory that maybe overall resource contention is causing problems.